[Exclusivity] Add analysis pass summarizing accesses to inout_aliasab…

devincoughlin opened the pull request:
…le params

Add an interprocedural SIL analysis pass that summarizes the accesses that
closures make on their @inout_aliasable captures. This will be used to
statically enforce exclusivity for calls to functions that take noescape
closures.

The analysis summarizes the accesses on each parameter independently and
uses the BottomUpIPAnalysis utility class to iterate to a fixed point when
there are cycles in the call graph.

For now, the analysis is not stored-property-sensitive -- that will come in a
latter commit.

devincoughlin committed:
[Exclusivity] Add analysis pass summarizing accesses to inout_aliasable args

Add an interprocedural SIL analysis pass that summarizes the accesses that
closures make on their @inout_aliasable captures. This will be used to
statically enforce exclusivity for calls to functions that take noescape
closures.

The analysis summarizes the accesses on each argument independently and
uses the BottomUpIPAnalysis utility class to iterate to a fixed point when
there are cycles in the call graph.

For now, the analysis is not stored-property-sensitive -- that will come in a
later commit.

devincoughlin commented:
\cc @rjmccall

atrick commented on lib/SILOptimizer/Analysis/AccessSummaryAnalysis.cpp:
```diff
@@ -0,0 +1,329 @@
+                                             SILFunctionArgument *parameter,
+                                             ParameterSummary &summary,
+                                             FunctionOrder &order) {
+  unsigned parameterIndex = parameter->getIndex();
```
What you're calling "parameter" is actually a SILFunctionArgument and
when you get its index you're getting the function argument index
(including indirect results), not the parameter index (user to index
into the function signature via SILFunctionType). This is a particular
pain point for me because I've spent several weeks worth of time
auditing and debugging code that doesn't make a proper distinction
between argument and parameter indices. For a given function
signature, argument indices change as SIL is lowered while parameter
indices do not.

I realize you only care about parameters here, so if you really want
to talk about parameters, then you can need to adjust the index
accordingly using
getFunction()->getConventions()->getSILArgIndexOfFirstParam(). But for
you, it's easier to just work with the SILArgument index and call it
what it is. If you were basing this on other code in SILOpt, point me
to it I'll cleanup that code too.

You're already doing the right thing to convert "applied argument
indices" to "function argument indices", although we lack a nice
interface for that.

atrick commented on lib/SILOptimizer/Analysis/AccessSummaryAnalysis.cpp:
```diff
@@ -0,0 +1,329 @@
+      // access markers.
+      break;
+    default:
+      llvm_unreachable("Unrecognized argument use");
```
Unreachable makes sense since enforcement isn't optional. However, a
hard assumption about the set of SIL operations that can consume an
address needs to be codified in the verifier. I made a note of this,
but I think at least a TODO would be nice.

atrick reviewed (approved):
Beautiful. I don't have any feedback on the design, just nits.

atrick commented on include/swift/SILOptimizer/Analysis/AccessSummaryAnalysis.h:
```diff
@@ -0,0 +1,195 @@
+  /// Maps functions to the information the analysis keeps for each function.
+  llvm::DenseMap<SILFunction *, FunctionInfo *> FunctionInfos;
+
+  llvm::SpecificBumpPtrAllocator<FunctionInfo> Allocator;
```
This is purely personal preference, but I have to say it drives me
crazy when member variables are hidden at the end of a class. Data
members are always the first thing I need to see and the thing I refer
back to frequently. I expect to find them listed together above the
constructor declaration.

devincoughlin commented:
Thanks Andy! I updated the patch use the correct terminology for arguments and move instance members to immediately precede constructors. I also removed several unused includes, added a TODO mentioning the need for a SILVerifier, and added a test for missing bodies when processing full applys.

devincoughlin commented:
@swift-ci Please smoke test and merge

atrick commented:
Awesome! Thanks @devincoughlin.

devincoughlin commented:
@swift-ci Please smoke test and merge
